### PR TITLE
Scope Promotion

### PR DESCRIPTION
**Pull request by erikh**
One thing tinyCI suffers from is a hard requirement on github tokens
with expansive permissions.

tinyCI now will ask for minimal permissions during the first oauth
cycle. If the user hits `uisvc:/login/upgrade`, they will be prompted to
increase their permissions.

Additionally, API calls that would otherwise perform operations that
need the requisite scope will redirect with status 302 instead,
requiring harder UI support for that behavior.

Finally, to indicate you need a scope, you can add the new
`x-token-scope` property to the swagger definitions to include a
handshake-level test before executing the handler.

Signed-off-by: Erik Hollensbe <github@hollensbe.org>

### CODE REVIEW

File: api/queuesvc/processors/processor.go [hunk 1]
```diff
@@ -374,7 +373,7 @@ func resolveParentInfo(h *handler.H, sub *types.Submission) (*types.Submission,
 		return nil, eErr
 	}
 
-	token := &oauth2.Token{}
+	token := &types.OAuthToken{}
```
Comment (capoferro) [inline api/queuesvc/processors/processor.go]: I know there's no import conflict in this file, but the inconsistency of `t` and `types` between files might catch me off guard.

File: handlers/handler.go [hunk 2]
```diff
@@ -29,12 +36,17 @@ const SessionUsername = "username"
...
+
 var (
 	errNoCapability  = errors.New("no capability to perform desired operation")
 	errInvalidCookie = errors.New("cookie was invalid")
 )
 
-var routeTransformer = regexp.MustCompile(`(?:{([^}]+)})+`)
+var (
+	routeTransformer = regexp.MustCompile(`(?:{([^}]+)})+`)
+)
```
Comment (capoferro) [inline handlers/handler.go]: Not a big deal, but this file has both `var foo` and `var ( foo )` declarations... maybe it's worth making these consistent or perhaps grouping the vars from 37 - 49 into one `var ( ... )` block?

File: handlers/handler.go [hunk 3]
```diff
@@ -54,8 +66,129 @@ type H struct {
...
+	if err != nil {
+		return nil, err
+	}
+
+	var name string
+	sess := h.Session(ctx)
+
+	username := sess.Get(SessionUsername)
+
+	// FIXME clean up this spaghetti -erikh
```
Comment (capoferro) [inline handlers/handler.go]: Explain what you intend to change (and perhaps leave a github ticket link in the comment with more details) so if we don't get back to this in the near future we know what you were thinking at the time.

File: handlers/handler.go (hunk 2 above)
Comment (erikh) [reply handlers/handler.go]: I tend to group private and public ones, but not together (as golint requirements make it look gory). I'll review this and make sure it's honoring that rules

File: handlers/handler.go (hunk 3 above)
Comment (erikh) [reply handlers/handler.go]: Yeah cristian complained about this the last time it cropped up. I'll make a ticket.

File: api/queuesvc/processors/processor.go (hunk 1 above)
Comment (erikh) [reply api/queuesvc/processors/processor.go]: there's two `types` packages in the system; in the case `t` is used both are in use.

File: handlers/handler.go (hunk 2 above)
Comment (erikh) [reply handlers/handler.go]: oh I see what I did here; I grouped up the errors. Ok, I'll collate these.

File: api/queuesvc/processors/processor.go (hunk 1 above)
Comment (capoferro) [reply api/queuesvc/processors/processor.go]: I understand your intention, I just mean I may find it strange if I'm flipping between files and have to remember that `types` refers to different packages depending on the file I'm in. Maybe it's ok, was a thought I had reading through.

**Pull request merged**